<a href="https://colab.research.google.com/github/anaisaoviedo-upb/Asociacion-para-canasta-de-mercado/blob/main/Reglas_de_asociacion_market.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Reglas de asociación - Canasta de mercado

* Preparación: variables binarias
* Aprendizaje: apriori
* Evaluación: soporte y confianza
* Interpretación: lectura de las reglas

In [ ]:
# Librerías básicas
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt


# 1. Preparación de datos

In [ ]:
data=pd.read_excel('market.xlsx',sheet_name=0)
data.head()

'Hair Conditioner_binarized' 'Lemons_binarized'  \
0                           No                 No   
1                           No                 No   
2                           No                 No   
3                           No                 No   
4                           No                 No   

  ' Standard coffee_binarized' ' Frozen Chicken Wings_binarized'  \
0                           No                                No   
1                           No                                No   
2                           No                                No   
3                           No                                No   
4                           No                                No   

  ' 98pct. Fat Free Hamburger_binarized' ' Sugar Cookies_binarized'  \
0                                     No                         No   
1                                     No                         No   
2                                     No                         No   
3                                     No                         No   
4                                     No                         No   

  ' Onions_binarized' ' Deli Ham_binarized' ' Dishwasher Detergent_binarized'  \
0                  No                    No                                No   
1                  No                    No                                No   
2                  No                    No                                No   
3                  No                    No                                No   
4                  No                    No                                No   

  ' Beets_binarized'  ... ' Lollipops_binarized'  \
0                 No  ...                     No   
1                 No  ...                     No   
2                 No  ...                     No   
3                 No  ...                     No   
4                 No  ...                     No   

  ' Plain White Bread_binarized' ' Blueberry Yogurt_binarized'  \
0                             No                            No   
1                             No                            No   
2                             No                            No   
3                             No                            No   
4                             No                            No   

  ' Frozen Chicken Thighs_binarized' ' Mixed Vegetables_binarized'  \
0                                 No                            No   
1                                 No                            No   
2                                 No                            No   
3                                 No                            No   
4                                 No                            No   

  ' Souring Pads_binarized' ' Tuna Spread_binarized'  \
0                        No                       No   
1                        No                       No   
2                        No                       No   
3                        No                       No   
4                        No                       No   

  ' Toilet Paper_binarized' ' White Wine_binarized'  \
0                        No                      No   
1                        No                      No   
2                        No                      No   
3                        No                      No   
4                        No                      No   

  ' Columbian Coffee_binarized'  
0                            No  
1                            No  
2                            No  
3                            No  
4                            No  

[5 rows x 255 columns]

In [ ]:
#Convertir a datos binarios
data=data.replace({'No': 0, 'Si': 1})
data.head()

/tmp/ipython-input-4-3192565404.py:2: FutureWarning: Downcasting behavior in `replace` is deprecated and will be removed in a future version. To retain the old behavior, explicitly call `result.infer_objects(copy=False)`. To opt-in to the future behavior, set `pd.set_option('future.no_silent_downcasting', True)`
  data=data.replace({'No': 0, 'Si': 1})


'Hair Conditioner_binarized'  'Lemons_binarized'  \
0                             0                   0   
1                             0                   0   
2                             0                   0   
3                             0                   0   
4                             0                   0   

   ' Standard coffee_binarized'  ' Frozen Chicken Wings_binarized'  \
0                             0                                  0   
1                             0                                  0   
2                             0                                  0   
3                             0                                  0   
4                             0                                  0   

   ' 98pct. Fat Free Hamburger_binarized'  ' Sugar Cookies_binarized'  \
0                                       0                           0   
1                                       0                           0   
2                                       0                           0   
3                                       0                           0   
4                                       0                           0   

   ' Onions_binarized'  ' Deli Ham_binarized'  \
0                    0                      0   
1                    0                      0   
2                    0                      0   
3                    0                      0   
4                    0                      0   

   ' Dishwasher Detergent_binarized'  ' Beets_binarized'  ...  \
0                                  0                   0  ...   
1                                  0                   0  ...   
2                                  0                   0  ...   
3                                  0                   0  ...   
4                                  0                   0  ...   

   ' Lollipops_binarized'  ' Plain White Bread_binarized'  \
0                       0                               0   
1                       0                               0   
2                       0                               0   
3                       0                               0   
4                       0                               0   

   ' Blueberry Yogurt_binarized'  ' Frozen Chicken Thighs_binarized'  \
0                              0                                   0   
1                              0                                   0   
2                              0                                   0   
3                              0                                   0   
4                              0                                   0   

   ' Mixed Vegetables_binarized'  ' Souring Pads_binarized'  \
0                              0                          0   
1                              0                          0   
2                              0                          0   
3                              0                          0   
4                              0                          0   

   ' Tuna Spread_binarized'  ' Toilet Paper_binarized'  \
0                         0                          0   
1                         0                          0   
2                         0                          0   
3                         0                          0   
4                         0                          0   

   ' White Wine_binarized'  ' Columbian Coffee_binarized'  
0                        0                              0  
1                        0                              0  
2                        0                              0  
3                        0                              0  
4                        0                              0  

[5 rows x 255 columns]

In [ ]:
data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1361 entries, 0 to 1360
Columns: 255 entries, 'Hair Conditioner_binarized' to ' Columbian Coffee_binarized'
dtypes: object(255)
memory usage: 2.6+ MB


In [ ]:
#Corrección del tipo de datos
data=data.astype('category')
data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1361 entries, 0 to 1360
Columns: 255 entries, 'Hair Conditioner_binarized' to ' Columbian Coffee_binarized'
dtypes: category(255)
memory usage: 369.9 KB


# 2. Aprendizaje - Evaluación



In [ ]:
from mlxtend.frequent_patterns import association_rules, apriori

#Paso 1 -Productos mas frecuentes: filtrar los que cumplen el soporte
frequent_itemsets = apriori(data, min_support = 0.05, max_len = 2, use_colnames=True)

frequent_itemsets

/usr/local/lib/python3.11/dist-packages/mlxtend/frequent_patterns/fpcommon.py:161: DeprecationWarning: DataFrames with non-bool types result in worse computationalperformance and their support might be discontinued in the future.Please use a DataFrame with bool type
  warnings.warn(


support                                           itemsets
0    0.05878                     ('Hair Conditioner_binarized')
1   0.093314           (' 98pct. Fat Free Hamburger_binarized')
2   0.055107                       (' Sugar Cookies_binarized')
3   0.080088                              (' Onions_binarized')
4   0.097722                        (' Potato Chips_binarized')
5   0.051433                             (' Waffles_binarized')
6   0.055841                               (' Plums_binarized')
7   0.054372                           (' Mushrooms_binarized')
8   0.056576                        (' Orange Juice_binarized')
9   0.050698                          (' Sour Cream_binarized')
10  0.069067            (' Pepperoni Pizza - Frozen_binarized')
11  0.092579                            (' Hot Dogs_binarized')
12  0.055841                       (' Frozen Shrimp_binarized')
13  0.060985                             (' Raisins_binarized')
14  0.057311                         (' Pancake Mix_binarized')
15  0.064658                             (' Bologna_binarized')
16  0.109478                          (' 2pct. Milk_binarized')
17  0.063924                        (' Popcorn Salt_binarized')
18  0.066128                            (' Tomatoes_binarized')
19  0.122704                                (' Eggs_binarized')
20   0.11903                         (' White Bread_binarized')
21  0.077884                                (' Cola_binarized')
22  0.053637                              (' Apples_binarized')
23  0.059515                             (' Bananas_binarized')
24  0.054372                       (' Sandwich Bags_binarized')
25  0.052902                    (' C Cell Batteries_binarized')
26   0.05878                        (' Hot Dog Buns_binarized')
27  0.055841                             (' Oranges_binarized')
28  0.079353                          (' Toothpaste_binarized')
29  0.054372                         (' Canned Tuna_binarized')
30  0.071271                      (' Hamburger Buns_binarized')
31   0.05878                        (' French Fries_binarized')
32  0.062454                       (' Ramen Noodles_binarized')
33  0.067597                       (' Domestic Beer_binarized')
34  0.057311                              (' Garlic_binarized')
35  0.077149                         (' Wheat Bread_binarized')
36  0.085231                        (' Sweet Relish_binarized')
37   0.07421                        (' Toilet Paper_binarized')
38  0.051433  (' White Bread_binarized', ' Potato Chips_bina...
39  0.052168       (' 2pct. Milk_binarized', ' Eggs_binarized')
40  0.051433  (' 2pct. Milk_binarized', ' White Bread_binari...
41  0.055107      (' Eggs_binarized', ' White Bread_binarized')

In [ ]:
# Paso 2 - Armar las reglas verificando confianza
rules = association_rules(frequent_itemsets, metric="confidence", min_threshold=0.4)
rules

antecedents                  consequents  \
0   (' White Bread_binarized')  (' Potato Chips_binarized')   
1  (' Potato Chips_binarized')   (' White Bread_binarized')   
2    (' 2pct. Milk_binarized')          (' Eggs_binarized')   
3          (' Eggs_binarized')    (' 2pct. Milk_binarized')   
4    (' 2pct. Milk_binarized')   (' White Bread_binarized')   
5   (' White Bread_binarized')    (' 2pct. Milk_binarized')   
6          (' Eggs_binarized')   (' White Bread_binarized')   
7   (' White Bread_binarized')          (' Eggs_binarized')   

   antecedent support  consequent support   support  confidence      lift  \
0            0.119030            0.097722  0.051433    0.432099  4.421702   
1            0.097722            0.119030  0.051433    0.526316  4.421702   
2            0.109478            0.122704  0.052168    0.476510  3.883414   
3            0.122704            0.109478  0.052168    0.425150  3.883414   
4            0.109478            0.119030  0.051433    0.469799  3.946889   
5            0.119030            0.109478  0.051433    0.432099  3.946889   
6            0.122704            0.119030  0.055107    0.449102  3.773010   
7            0.119030            0.122704  0.055107    0.462963  3.773010   

   representativity  leverage  conviction  zhangs_metric   jaccard  certainty  \
0               1.0  0.039801    1.588793       0.878399  0.311111   0.370592   
1               1.0  0.039801    1.859825       0.857655  0.311111   0.462315   
2               1.0  0.038734    1.675861       0.833775  0.289796   0.403292   
3               1.0  0.038734    1.549137       0.846344  0.289796   0.354479   
4               1.0  0.038402    1.661576       0.838425  0.290456   0.398162   
5               1.0  0.038402    1.568093       0.847516  0.290456   0.362283   
6               1.0  0.040501    1.599152       0.837755  0.295276   0.374669   
7               1.0  0.040501    1.633586       0.834262  0.295276   0.387850   

   kulczynski  
0    0.479207  
1    0.479207  
2    0.450830  
3    0.450830  
4    0.450949  
5    0.450949  
6    0.456032  
7    0.456032

In [ ]:
rules= rules.sort_values("confidence",ascending=False)
rules

antecedents                  consequents  \
1  (' Potato Chips_binarized')   (' White Bread_binarized')   
2    (' 2pct. Milk_binarized')          (' Eggs_binarized')   
4    (' 2pct. Milk_binarized')   (' White Bread_binarized')   
7   (' White Bread_binarized')          (' Eggs_binarized')   
6          (' Eggs_binarized')   (' White Bread_binarized')   
0   (' White Bread_binarized')  (' Potato Chips_binarized')   
5   (' White Bread_binarized')    (' 2pct. Milk_binarized')   
3          (' Eggs_binarized')    (' 2pct. Milk_binarized')   

   antecedent support  consequent support   support  confidence      lift  \
1            0.097722            0.119030  0.051433    0.526316  4.421702   
2            0.109478            0.122704  0.052168    0.476510  3.883414   
4            0.109478            0.119030  0.051433    0.469799  3.946889   
7            0.119030            0.122704  0.055107    0.462963  3.773010   
6            0.122704            0.119030  0.055107    0.449102  3.773010   
0            0.119030            0.097722  0.051433    0.432099  4.421702   
5            0.119030            0.109478  0.051433    0.432099  3.946889   
3            0.122704            0.109478  0.052168    0.425150  3.883414   

   representativity  leverage  conviction  zhangs_metric   jaccard  certainty  \
1               1.0  0.039801    1.859825       0.857655  0.311111   0.462315   
2               1.0  0.038734    1.675861       0.833775  0.289796   0.403292   
4               1.0  0.038402    1.661576       0.838425  0.290456   0.398162   
7               1.0  0.040501    1.633586       0.834262  0.295276   0.387850   
6               1.0  0.040501    1.599152       0.837755  0.295276   0.374669   
0               1.0  0.039801    1.588793       0.878399  0.311111   0.370592   
5               1.0  0.038402    1.568093       0.847516  0.290456   0.362283   
3               1.0  0.038734    1.549137       0.846344  0.289796   0.354479   

   kulczynski  
1    0.479207  
2    0.450830  
4    0.450949  
7    0.456032  
6    0.456032  
0    0.479207  
5    0.450949  
3    0.450830

In [ ]:
rules.to_excel('./resultados.xlsx')

# **INTERPRETACIÓN**

Regla 1. De los que compran huevos, el 42.5% compran también leche